### DISCUSSION TITLE
Add side-effects to services

### DISCUSSION THREAD

**Discussion by nicospz** [statelyai/xstate#4066] [Ideas]
[This](https://github.com/statelyai/xstate/issues/547) was posted 3 years ago and it's an excellent way to add side-effects to a live service.
I wonder if we are going to get an official API for this. Specially since I have problems typing this function.
It should check that `actionType` is a valid key for the machine actions, but even if I make it a generic and pass the machine type I can't get it to work with `MachineOptionsFrom`. Am I missing something or is it not possible to get the action keys like this?

**Comment by davidkpiano** [comment]
Can you describe your use-case further?

cc. @Andarist for a potential `state.actions` use-case (this property will be removed in v5)
